tests/phpunit/api/v3/ContributionTest.php

```diff
@@ -2523,6 +2523,47 @@ public function testRepeatTransactionTestRecurId() {
     $this->assertEquals($contributionRecur['values'][1]['is_test'], $repeatedContribution['values'][2]['is_test']);
   }
 
+  /**
+   * Test repeat contribution accepts recur_id instead of
+   * original_contribution_id.
+   *
+   * @throws \CRM_Core_Exception
+   */
+  public function testRepeatTransactionPreviousContributionRefunded(): void {
+    $contributionRecur = $this->callAPISuccess('contribution_recur', 'create', [
+      'contact_id' => $this->_individualId,
+      'installments' => '12',
+      'frequency_interval' => '1',
+      'amount' => '100',
+      'contribution_status_id' => 1,
+      'start_date' => '2012-01-01 00:00:00',
+      'currency' => 'USD',
+      'frequency_unit' => 'month',
+      'payment_processor_id' => $this->paymentProcessorID,
+    ]);
+    $this->callAPISuccess('contribution', 'create', array_merge(
+        $this->_params,
+        [
+          'contribution_recur_id' => $contributionRecur['id'],
+          'contribution_status_id' => 'Refunded',
+        ]
+      )
+    );
+
+    $this->callAPISuccess('contribution', 'repeattransaction', [
+      'contribution_recur_id' => $contributionRecur['id'],
+      'trxn_id' => 1234,
+    ]);
+    $contributions = $this->callAPISuccess('contribution', 'get', [
+      'contribution_recur_id' => $contributionRecur['id'],
+      'sequential' => 1,
+    ]);
+    // We should have contribution 0 in "Refunded" status and contribution 1 in "Pending" status
+    $this->assertEquals(2, $contributions['count']);
+    $this->assertEquals(7, $contributions['values'][0]['contribution_status_id']);
+    $this->assertEquals(2, $contributions['values'][1]['contribution_status_id']);
+  }
+
```

Comment:
@eileenmcnaughton @mattwire I am happy with the logic, but I am not sure moving forward, are we relying on APIv4 to do CRUD operations on new unit tests? if it doesn't matter unless the UT captures the fix/change accurately then I am ok here.

Reply:
@monishdeb I think you are saying 'would this be better if we were using apiv4 instead of 3 in the tests.

I guess we are moving that way - but we are not so far down the track I'm alarmed at adding new tests that use apiv3 - in this case, however, I think v4 would be better on the final get because it would make it easy to check contribution_status_id:name rather than the hard-coded '2' & '7'

Reply:
yes `contribution_status_id:name` is the other reason I felt like APIv4 would be appropriate here.

( I wish there's a emoticon for hi5 <img src="https://user-images.githubusercontent.com/3735621/151103269-b6a68999-92cd-408c-9406-623b47f72df0.jpeg" data-canonical-src="https://user-images.githubusercontent.com/3735621/151103269-b6a68999-92cd-408c-9406-623b47f72df0.jpeg" width="30" height="30" /> )

Reply:
![five](https://c.tenor.com/9DH2mkJLXIwAAAAM/highf-fve-30rock.gif)

Reply:
@eileenmcnaughton @monishdeb I agree with preferring to use API4 but I'm not sure I understand fully how we would implement it for the tests? We can set `$this->_apiversion = 4;` but in this case repeattransaction does not exist for API4 so we would have to change api version during the test which feels a bit messy?

Reply:
@mattwire generally I just call the v4 api direct in the tests (unless specifically testing apiv3 or both 3 & 4)  - the `callApiSuccess` helpers are helpful for apiv3 but don't add much to api v4 - so it's really just that last Contribution::get where using apiv4 would make it easy to eliminate the checks on hard-coded status ids (which it the bit that raised the discussion)